README.md
<!--
 * @Author       : liqiao
 * @Date         : 2023-05-05 14:33:46
 * @LastEditors  : liqiao
 * @LastEditTime : 2023-05-10 15:40:28
 * @Description  : Do not edit
 * @FilePath     : /code-shortcut-generation/README.md
-->

## 特征

项目中往往会有一些使用频率很高的公共组件，比如管理后台的表格、表单等等，每次都需要粘贴复制，此插件的作用是用于快速生成公共代码，不再手动处理；

## 使用说明

原理是复制需要的模版代码，生成到当前文件中； 

以code开头，共有三个参数，第一个参数为需要读取的文件绝对路径，第二个参数为原始组件名称，第三个参数为当前生成的组件名称；

代码提示：
```
code 读取文件绝对路径>读取文件组件名称>当前组件名称>
```

<!-- \!\[feature X\]\(images/pikachu.png\) -->

> Tip: 可以整理出需要自动生成的多个代码文件，放在同一个文件夹下进行管理；


CHANGELOG.md
# Change Log

All notable changes to the "code-shortcut-generation" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

script/test.js
/*
 * @Author       : liqiao
 * @Date         : 2023-05-09 14:45:55
 * @LastEditors  : liqiao
 * @LastEditTime : 2023-05-09 14:45:56
 * @Description  : Do not edit
 * @FilePath     : /code-shortcut-generation/src/script/test.ts
 */
console.log(111);


src/utils/util.ts
/*
 * @Author       : liqiao
 * @Date         : 2023-05-10 14:01:08
 * @LastEditors  : liqiao
 * @LastEditTime : 2023-05-10 14:06:39
 * @Description  : Do not edit
 * @FilePath     : /code-shortcut-generation/src/utils/util.ts
 */
import * as vscode from "vscode";
const fs = require("fs");
const path = require("path");

/**
 * 获取当前所在工程根目录，有3种使用方法：<br>
 * getProjectPath(uri) uri 表示工程内某个文件的路径<br>
 * getProjectPath(document) document 表示当前被打开的文件document对象<br>
 * getProjectPath() 会自动从 activeTextEditor 拿document对象，如果没有拿到则报错
 * @param {*} document
 */
export const getProjectPath = (document?: any) => {
  if (!document) {
    document = vscode.window.activeTextEditor
      ? vscode.window.activeTextEditor.document
      : null;
  }
  if (!document) {
    // this.showError('当前激活的编辑器不是文件或者没有文件被打开！');
    return "";
  }
  const currentFile = (document.uri ? document.uri : document).fsPath;
  let projectPath = null;

  let workspaceFolders = vscode.workspace?.workspaceFolders?.map(
    (item) => item.uri.path
  );
  // 由于存在Multi-root工作区，暂时没有特别好的判断方法，先这样粗暴判断
  // 如果发现只有一个根文件夹，读取其子文件夹作为 workspaceFolders
  if (
    workspaceFolders?.length === 1 &&
    workspaceFolders[0] === vscode.workspace.rootPath
  ) {
    const rootPath = workspaceFolders[0];
    const files = fs.readdirSync(rootPath);

    console.log("🚀 ~ file: util.ts:44 ~ getProjectPath ~ files:", files);

    workspaceFolders = files
      .filter((name: any) => !/^\./g.test(name))
      .map((name: any) => path.resolve(rootPath, name));
  }

  console.log(
    "🚀 ~ file: util.ts:47 ~ getProjectPath ~ workspaceFolders:",
    workspaceFolders
  );

  workspaceFolders?.forEach((folder) => {
    if (currentFile.indexOf(folder) === 0) {
      projectPath = folder;
    }
  });

  console.log(
    "🚀 ~ file: util.ts:45 ~ getProjectPath ~ projectPath:",
    projectPath
  );

  if (!projectPath) {
    // this.showError('获取工程根路径异常！');
    return "";
  }
  return projectPath;
};


script/copyCode.js
console.log("test");

/**
 * 
 * @param  {String} readerUrl  读取内容文件路径，绝对路径
 * @param  {String} writerUrl  写入文件路径，绝对路径
 * @param  {String} replaceTemplateName  需要替换的模版名称
 */

const fs = require("fs");
const path = require("path");
let readerFilePathData = "";

// 创建一个可以写入的流，写入到指定文件中，
const writerOperate = () => {
  const writerFilePath = path.resolve(__dirname, "../src/test.tsx");
  console.log("🚀 ~ file: copyCode.js:18 ~ writerFilePath:", writerFilePath);


  const writerStream = fs.createWriteStream(writerFilePath);

  // 使用 utf8 编码写入数据
  console.log(
    "🚀 ~ file: copyCode.js:43 ~ readerFilePathData:",
    readerFilePathData
  );

  const readerFilePathDataCurr = readerFilePathData.replace(/TemplateCode/g, "TestTemplateCode");

  writerStream.write(readerFilePathDataCurr, "UTF8");

  // 标记文件末尾
  writerStream.end();

  // 处理流事件 --> data, end, and error
  writerStream.on("finish", function () {
    console.log("写入完成。");
  });

  writerStream.on("error", function (err) {
    console.log(err.stack);
  });

  console.log("写入程序执行完毕");
};

// 获取文件模版内容
const readerOperate = () => {
  const readerFilePath = path.resolve(
    __dirname,
    "../src/codeTemplates/modal.tsx"
  );
  // console.log("🚀 ~ file: copyCode.js:18 ~ readerFilePath:", readerFilePath);
  // 创建可读流
  // const readerStream = fs.createReadStream('input.txt');
  const readerStream = fs.createReadStream(readerFilePath);

  // 设置编码为 utf8。
  readerStream.setEncoding("UTF8");

  // 处理流事件 --> data, end, and error
  readerStream.on("data", function (chunk) {
    readerFilePathData += chunk;
  });

  // 文件读取成功
  readerStream.on("end", function () {
    // console.log("readerFilePathData-----end");
    writerOperate();
  });

  readerStream.on("error", function (err) {
    console.log(err.stack);
  });

  // console.log("读取程序执行完毕");
};

readerOperate();


script/deleteFile.js
/*
 * @Author       : liqiao
 * @Date         : 2023-05-09 14:44:48
 * @LastEditors  : liqiao
 * @LastEditTime : 2023-05-09 15:21:27
 * @Description  : Do not edit
 * @FilePath     : /code-shortcut-generation/script/deleteFile.js
 */

const fs = require("fs");
const path = require("path");

/**
 * 删除某一个包下面的需要符合格式的文件。
 * @param  {String} url  文件路径，绝对路径
 * @param  {String} name 需要删除的文件名称
 * @return {Null}
 */

const deleteFile = (url, name) => {
  let files = [];
  //判断给定的路径是否存在
  if (fs.existsSync(url)) {
    files = fs.readdirSync(url); //返回文件和子目录的数组
    // console.log("🚀 ~ file: deleteFile.ts:25 ~ deleteFile ~ files:", files);

    files.forEach((file, index) => {
      // console.log("🚀 ~ file: deleteFile.ts:38 ~ files.forEach ~ file:", file);

      var curPath = path.join(url, file);
      if (fs.statSync(curPath).isDirectory()) {
        //同步读取文件夹文件，如果是文件夹，则函数回调
        deleteFile(curPath, name);
      } else {
        if (file.indexOf(name) > -1) {
          fs.unlinkSync(curPath);
          // console.log("删除文件：" + curPath);
        }
      }
    });
  } else {
    console.log("给定的路径不存在！");
  }
};

// 传入script文件夹名称
deleteFile("../", "code-shortcut-generation");


src/extension_start.ts
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
import * as vscode from "vscode";

// This method is called when your extension is activated
// Your extension is activated the very first time the command is executed
export function activate(context: vscode.ExtensionContext) {
  // Use the console to output diagnostic information (console.log) and errors (console.error)
  // This line of code will only be executed once when your extension is activated
  console.log(
    'Congratulations, your extension "code-shortcut-generation" is now active!'
  );

  // The command has been defined in the package.json file
  // Now provide the implementation of the command with registerCommand
  // The commandId parameter must match the command field in package.json
  let disposable = vscode.commands.registerCommand(
    "code-shortcut-generation.helloWorld",
    () => {
      // The code you place here will be executed every time your command is executed
      // Display a message box to the user
      vscode.window.showInformationMessage(
        "Hello World from code-shortcut-generation!"
      );
    }
  );

  context.subscriptions.push(disposable);
}

// This method is called when your extension is deactivated
export function deactivate() {}


script/copyCode_start.js
console.log("test");

var fs = require("fs");
const path = require("path");

const readerFilePath = path.resolve(
  __dirname,
  "../src/codeTemplates/modal.tsx"
);
// console.log("🚀 ~ file: copyCode.js:18 ~ readerFilePath:", readerFilePath);
let readerFilePathData = "";
// 创建可读流
// var readerStream = fs.createReadStream('input.txt');
var readerStream = fs.createReadStream(readerFilePath);

// 设置编码为 utf8。
readerStream.setEncoding("UTF8");

// 处理流事件 --> data, end, and error
readerStream.on("data", function (chunk) {
  readerFilePathData += chunk;
});

readerStream.on("end", function () {
  console.log("readerFilePathData-----end");
  // 创建一个可以写入的流，写入到指定文件中

  const writerFilePath = path.resolve(__dirname, "../src/test.tsx");
  console.log("🚀 ~ file: copyCode.js:18 ~ writerFilePath:", writerFilePath);
  var writerStream = fs.createWriteStream(writerFilePath);

  // 使用 utf8 编码写入数据
  console.log(
    "🚀 ~ file: copyCode.js:43 ~ readerFilePathData:",
    readerFilePathData
  );

  writerStream.write(readerFilePathData, "UTF8");

  // 标记文件末尾
  writerStream.end();

  // 处理流事件 --> data, end, and error
  writerStream.on("finish", function () {
    console.log("写入完成。");
  });

  writerStream.on("error", function (err) {
    console.log(err.stack);
  });

  console.log("写入程序执行完毕");
});

readerStream.on("error", function (err) {
  console.log(err.stack);
});

console.log("读取程序执行完毕");


script/update-version.js
/*
 * @Author       : liqiao
 * @Date         : 2023-05-09 15:40:30
 * @LastEditors  : liqiao
 * @LastEditTime : 2023-05-09 15:49:01
 * @Description  : Do not edit
 * @FilePath     : /code-shortcut-generation/script/update-version.js
 */
const path = require('path');
const fs = require('fs');
const packageJson = require('../package.json');

const filePath = path.resolve(__dirname, '../package.json');

function getJsonData() {
    return packageJson;
}

function getNewVersion(version) {
    const [major = 1, minor = 0, patch = 0] = version.split('.');
    let newPatch = +patch + 1;
    let newMinor = +minor;
    let newMajor = +major;
    if (newPatch > 99) {
        newPatch = 0;
        newMinor += 1;
    }
    if (newMinor > 99) {
        newMinor = 0;
        newMajor += 1;
    }
    return `${newMajor}.${newMinor}.${newPatch}`;
}

function updateVersion(data) {
    console.log("🚀 ~ file: update-version.js:37 ~ updateVersion ~ filePath, data:", filePath, data);

    fs.writeFileSync(filePath, data);
}
function exec() {
    const jsonData = getJsonData();
    console.log("🚀 ~ file: update-version.js:40 ~ exec ~ jsonData:", jsonData.version);
    const newVersion = getNewVersion(jsonData.version);
    console.log("🚀 ~ file: update-version.js:42 ~ exec ~ newVersion:", newVersion);
    jsonData.version = newVersion;
    console.log("🚀 ~ file: update-version.js:43 ~ exec ~ jsonData:", jsonData.version);
    updateVersion(JSON.stringify(jsonData, null, 4));
}

exec();

src/extension_useState.ts
/*
 * @Author       : liqiao
 * @Date         : 2023-05-05 17:51:29
 * @LastEditors  : liqiao
 * @LastEditTime : 2023-05-06 17:30:02
 * @Description  : Do not edit
 * @FilePath     : /code-shortcut-generation/src/extension.ts
 */

import * as vscode from "vscode";

const WRAPPING_TRIGGER_CHARACTERS = [
  "{",
  "}",
  "[",
  "]",
  "(",
  ")",
  "'",
  '"',
] as const;
const TRIGGER_CHARACTERS = [
  ...WRAPPING_TRIGGER_CHARACTERS,
  "/",
  "0",
  "1",
  "2",
  "3",
  "4",
  "5",
  "6",
  "7",
  "8",
  "9",
] as const;

const useStateCommandId = "vscode-pikaChu-useState";

function genValue(text: string) {
  console.log("🚀 ~ file: extension.ts:37 ~ genValue ~ text:", text);
  const [startChar, endChar] =
    text.length === 1 ? [text[0], null] : [text[0], text[text.length - 1]];
  console.log(
    "🚀 ~ file: extension.ts:39 ~ genValue ~ startChar, endChar:",
    startChar,
    endChar
  );

  if (
    !(WRAPPING_TRIGGER_CHARACTERS as readonly string[]).includes(startChar) ||
    startChar === endChar
  ) {
    return text;
  }

  switch (startChar) {
    case '"':
    case "'":
      return text + startChar;
    case "{":
      return text + "}";
    case "[":
      return text + "]";
    case "(":
      return text + ")";

    default:
      return text;
  }
}

class MyCompletionItemProvider implements vscode.CompletionItemProvider {
  private position?: vscode.Position;
  private str = "";

  constructor() {}

  // provideCompletionItems 接受 vscode 传入的当前 文档(document) 和 当前光标位置(position) 两个参数，
  // 通过这两个参数得到当前行输入的文本，对文本进一步解析，生成我们想要的代码片段，
  // 然后通过 new vscode.CompletionItem 生成一个代码提示实例，return 出去
  public provideCompletionItems(
    document: vscode.TextDocument,
    position: vscode.Position,
    token: vscode.CancellationToken,
    context: vscode.CompletionContext
  ): vscode.ProviderResult<
    vscode.CompletionItem[] | vscode.CompletionList<vscode.CompletionItem>
  > {
    this.position = position;

    // 返回由行号表示的文本行
    const linePrefix = document
      .lineAt(position)
      .text.slice(0, position.character);

    // 如果文本前缀不为 const ，则不做提示
    if (!linePrefix?.startsWith("const ") || !linePrefix.split("const ")[1]) {
      this.str = "";
      return [];
    }

    const [name, type, value] = linePrefix.split("const ")[1].split("/");

    this.str =
      "const [" +
      name +
      ", set" +
      name.replace(name[0], name[0].toUpperCase()) +
      "] = useState";

    // ts类型
    if (type) {
      if (!value) {
        this.str += `(${genValue(type)})`;
      } else {
        if (type === "<" || type === "<>") {
          this.str += "<>";
        } else {
          this.str += `<${type}>`;
        }
      }
    }

    if (value) {
      this.str += `(${genValue(value)})`;
    } else if (!type) {
      this.str += "()";
    }

    const snippetCompletion = new vscode.CompletionItem(
      linePrefix,
      // vscode.CompletionItemKind.Snippet
      vscode.CompletionItemKind.Color
    );

    snippetCompletion.documentation = this.str; // 一个人类可读的字符串，表示文档注释。
    snippetCompletion.detail = "enter / split code"; // 一个人类可读的字符串，包含关于该项的附加信息，如类型或符号信息。
    snippetCompletion.range = new vscode.Range(
      new vscode.Position(position.line, position.character),
      new vscode.Position(position.line, position.character)
    );

    snippetCompletion.filterText = "";

    return [snippetCompletion];
  }

  // 光标选中当前自动补全item时触发动作
  public resolveCompletionItem(
    item: vscode.CompletionItem,
    token: vscode.CancellationToken
  ): vscode.ProviderResult<vscode.CompletionItem> {
    const label = item.label;
    if (this.position && typeof label === "string") {
      item.command = {
        command: useStateCommandId,
        title: "refactor",
        arguments: [this.position.translate(0, label.length + 1), this.str], // 这里可以传递参数给该命令
      };
    }

    return item;
  }
}

export function activate(context: vscode.ExtensionContext) {
  console.log(
    'Congratulations, your extension "code-shortcut-generation" is now active!'
  );

  const commandHandler = (
    editor: vscode.TextEditor,
    edit: vscode.TextEditorEdit,
    position: vscode.Position,
    str: string
  ) => {
    const lineText = editor.document.lineAt(position.line).text;
    const startSpaces = lineText.length - lineText.trimStart().length;

    edit.delete(
      new vscode.Range(
        position.with(undefined, startSpaces),
        position.with(undefined, lineText.length)
      )
    );

    edit.insert(position.with(undefined, startSpaces), str);

    return Promise.resolve([]);
  };

  context.subscriptions.push(
    vscode.commands.registerTextEditorCommand(useStateCommandId, commandHandler)
  );

  const disposable = vscode.languages.registerCompletionItemProvider(
    [
      "javascript",
      "javascriptreact",
      "typescript",
      "typescriptreact",
      "vue",
      "html",
    ],
    new MyCompletionItemProvider(),
    ...TRIGGER_CHARACTERS
  );

  context.subscriptions.push(disposable);
}

// This method is called when your extension is deactivated
export function deactivate() {}


src/extension.ts
/*
 * @Author       : liqiao
 * @Date         : 2023-05-09 16:06:57
 * @LastEditors  : liqiao
 * @LastEditTime : 2023-05-10 15:40:50
 * @Description  : Do not edit
 * @FilePath     : /code-shortcut-generation/src/extension.ts
 */

import * as vscode from "vscode";

const fs = require("fs");
const path = require("path");

// 分隔标志符
const separatorIdentifier = [">"] as const;

const useStateCommandId = "vscode-pikaChu-code-generation";

class MyCompletionItemProvider implements vscode.CompletionItemProvider {
  private position?: vscode.Position;
  private str = "";

  constructor() {}

  // provideCompletionItems 接受 vscode 传入的当前 文档(document) 和 当前光标位置(position) 两个参数，
  // 通过这两个参数得到当前行输入的文本，对文本进一步解析，生成我们想要的代码片段，
  // 然后通过 new vscode.CompletionItem 生成一个代码提示实例，return 出去
  public provideCompletionItems(
    document: vscode.TextDocument,
    position: vscode.Position,
    token: vscode.CancellationToken,
    context: vscode.CompletionContext
  ): vscode.ProviderResult<
    vscode.CompletionItem[] | vscode.CompletionList<vscode.CompletionItem>
  > {
    this.position = position;

    // 返回由行号表示的文本行
    const linePrefix = document
      .lineAt(position)
      .text.slice(0, position.character);

    // 如果文本前缀不为 code ，则不做提示
    if (!linePrefix?.startsWith("code ") || !linePrefix.split("code ")[1]) {
      this.str = "";
      return [];
    }

    /**
     * readerUrl: 读取文件路径
     * originalValue: 原始字符串
     * replaceValue: 替换字符串
     */
    const [readerUrl, originalValue, replaceValue] = linePrefix
      .split("code ")[1]
      .split(">");

    const readerFilePath = path.resolve(__dirname, readerUrl);

    const readerFile = fs.readFileSync(readerFilePath, "utf-8");

    const readerFileCurr = readerFile.replaceAll(originalValue, replaceValue);

    this.str = readerFileCurr;

    const snippetCompletion = new vscode.CompletionItem(
      linePrefix,
      vscode.CompletionItemKind.Snippet
      // vscode.CompletionItemKind.Color
    );

    snippetCompletion.documentation = `Create a file template：${
      replaceValue ?? ""
    }`; // 一个人类可读的字符串，表示文档注释。
    snippetCompletion.detail = "enter > split code"; // 一个人类可读的字符串，包含关于该项的附加信息，如类型或符号信息。

    snippetCompletion.range = new vscode.Range(
      new vscode.Position(position.line, position.character),
      new vscode.Position(position.line, position.character)
    );

    snippetCompletion.filterText = "";

    return [snippetCompletion];
  }

  // 光标选中当前自动补全item时触发动作
  public resolveCompletionItem(
    item: vscode.CompletionItem,
    token: vscode.CancellationToken
  ): vscode.ProviderResult<vscode.CompletionItem> {
    const label = item.label;
    if (this.position && typeof label === "string") {
      item.command = {
        command: useStateCommandId,
        title: "refactor",
        arguments: [this.position.translate(0, label.length + 1), this.str], // 这里可以传递参数给该命令
      };
    }

    return item;
  }
}

export function activate(context: vscode.ExtensionContext) {
  console.log(
    'Congratulations, your extension "code-shortcut-generation" is now active!'
  );

  const commandHandler = (
    editor: vscode.TextEditor,
    edit: vscode.TextEditorEdit,
    position: vscode.Position,
    str: string
  ) => {
    const lineText = editor.document.lineAt(position.line).text;
    const startSpaces = lineText.length - lineText.trimStart().length;

    edit.delete(
      new vscode.Range(
        position.with(undefined, startSpaces),
        position.with(undefined, lineText.length)
      )
    );

    edit.insert(position.with(undefined, startSpaces), str);

    return Promise.resolve([]);
  };

  context.subscriptions.push(
    vscode.commands.registerTextEditorCommand(useStateCommandId, commandHandler)
  );

  const disposable = vscode.languages.registerCompletionItemProvider(
    [
      "javascript",
      "javascriptreact",
      "typescript",
      "typescriptreact",
      "vue",
      "html",
    ],
    new MyCompletionItemProvider(),
    ...separatorIdentifier
  );

  context.subscriptions.push(disposable);
}

export function deactivate() {}
